app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt

```diff
@@ -2,17 +2,182 @@ package compiler.semantic_analysis
+                checkExpression(expression.rightOperand)
+            }
+            is Expression.Conditional -> {
+                checkExpression(expression.condition)
+                checkExpression(expression.resultWhenTrue)
+                checkExpression(expression.resultWhenFalse)
+            }
+        }
+    }
+
+    private fun checkVariable(variable: Variable) {
+        if (variable.value != null) {
+            checkExpression(variable.value)
+        }
+    }
+
+    private fun checkFunctionCalls(program: Program) {
+        fun checkFunction(function: Function) {
+            fun checkBlock(block: StatementBlock) {
+                for (statement in block) {
+                    when (statement) {
+                        is Statement.Evaluation -> checkExpression(statement.expression)
+                        is Statement.VariableDefinition -> checkVariable(statement.variable)
+                        is Statement.FunctionDefinition -> checkFunction(statement.function)
+                        is Statement.Assignment -> checkExpression(statement.value)
+                        is Statement.Block -> checkBlock(statement.block)
+                        is Statement.Conditional -> {
+                            checkExpression(statement.condition)
+                            checkBlock(statement.actionWhenTrue)
+                            statement.actionWhenFalse?.let { checkBlock(it) }
+                        }
+                        is Statement.Loop -> {
+                            checkExpression(statement.condition)
+                            checkBlock(statement.action)
+                        }
+                        is Statement.FunctionReturn -> checkExpression(statement.value)
+                    }
+                }
+            }
+
+            checkBlock(function.body)
+        }
+
+        program.globals.forEach { if (it is Program.Global.FunctionDefinition) checkFunction(it.function) }
+    }
+
+    private fun checkFunctionDefinitions(program: Program) {
+        fun defaultParametersBeforeNonDefault(function: Function): Boolean {
+            var encounteredDefault = false
+            for (parameter in function.parameters) {
+                if (parameter.defaultValue == null) {
+                    if (encounteredDefault)
+                        return true
+                } else {
+                    encounteredDefault = true
+                }
+            }
+            return false
+        }
+
+        fun checkFunction(function: Function, global: Boolean) {
+            if (defaultParametersBeforeNonDefault(function)) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.DefaultParametersNotLast(function))
+            }
+
+            // TODO: generate variables for default parameters
+
+            fun checkBlock(block: StatementBlock) {
+                for (statement in block) {
+                    when (statement) {
```

Comment:
![image](https://user-images.githubusercontent.com/59764862/202900324-6733d560-0d51-488a-adb0-4f5a84c9419e.png)
Maybe worth adding `else -> {}` on the bottom